### ISSUE TITLE
makePreloader Documentation

### ISSUE THREAD

**Issue by nickarora** [outlandishideas/kasia#24]
Is there any additional documentation or examples available to show how to use makePreloader for the purpose of server-side rendering?  The current Readme is very sparse.

**Comment by sdgluck** [comment]
The documentation is very sparse indeed! We're sorry about that, other things have come first on the list of priorities, but we haven't forgotten. The aim is to have a boilerplate that can demonstrate these features of the library ASAP.

Essentially the steps to follow would be something like:

- Collect all components against the route.
- Map the components to the result of calling their `makePreloader` function, if they have one, passing in the props. Each call to `makePreloader` returns an array of tuples. Each tuple consists of a saga and an action object to call it with.
- Call each saga with its corresponding action objects __in the same order__ as they were generated.

If you have the time and inclination to put in a PR for some extra documentation, it would be much appreciated. :+1:

**Comment by nickarora** [comment]
I wouldn't mind helping, however I want to ensure that I've set things up correctly before attempting to document anything.  Currently I've got all of the above steps up and running, however I'm noticing a problem.

On the server, I'm doing the following:
1. I collect all components against the route using renderProps.components
2. Map the components to the results of calling their makePreloader function.
3. I run the resulting fetch/action tuples
4. Once the sagas have been run, I render the results to string -- then send back response to the client.

This results in a perfectly rendered page with no wait because the data has been pre-rendered.

However, once my client side JS takes over, my connectWpPost component _thinks_ it does not have available data, despite the store being pre-populated with the fetched wordpress data (query/entity).  I believe this is because the client side component still doesn't have a queryId, so it is unable to figure out that the data in the store belongs to it.  

Is there some way around this?  Any insight would be appreciated.

**Comment by nickarora** [comment]
As an update, I was able to debug the issue above.  

`makePreloader` assigns a queryId to the intermediate components when it renders server-side.  The first time makePreloader is run it begins by assigning a queryId of`'0'`.  With each refresh subsequent run of makePreloader, it incrementally updates the queryId it assigns (`'1'`, `'2'`, etc.).  

Because idGen is a singleton, this number persists and will continue to grow -- even between different requests.  If its your 5th reload and you only have one kasia-connected component, then Kasia will populate wordpress.query with `'5': { ... }`

This becomes a problem on the client side.  Because the client side script is running for the first time, it will look for a query with the key `'0'`.  If none is found (because `'5'` exists instead), it will run out to fetch the data -- and ignore the fact that the data already exists in the store.

I was able to fix this by changing lines 81/82 of connectWpPost.js from:

```
const id = idGen()
queryIds.push(id)
```

to 

```
const id = queryIds.length
queryIds.push(id)
```

**Comment by sdgluck** [comment]
Hey @nickarora, thanks for your time on this. Has the change you shared meant that server-side rendering is working for you now?

Great that you'd like help on this! Let's make this a place for discussion on the server-side rendering solution. I think perhaps not relying on an array that sits outside of the store would be a start - for example, maintaining the `queryIds` within the store.

**Comment by bt-frog** [comment]
@nickarora - Thank you for posting this; I was just about to post something similar. I also appreciate the library too. Would you mind sharing what your fetchComponentData (or whatever you're calling it on the server) looks like?

**Comment by sdgluck** [comment]
@bt-frog From the comments above it looks like there is a bug in the library that will prevent anyone from using server-side rendering as it currently stands. If @nickarora's solution above has meant that it now works for him, I will ask if he can submit a PR with this change.

Just quickly, on the example you have shared, what is `params`? The function returned by `makePreloader` doesn't accept any arguments. If `params` is the render props object then creation of the preloader should be amended to pass them in to `makePreloader` and not the preloader itself:

    const preloader = component.makePreloader(params)()

**Comment by bt-frog** [comment]
Hi Sam,

Thanks for the update to the makePreloader call; while powering through the server-side rendering issues, I left that one there. The `params` argument is from the redux-router/server match `routerState.params` callback:
```
store.dispatch(match(url, (error, redirectLocation, routerState) => {
    …
      // res.status(200).end(getMarkup(store));
      fetchComponentData(routerState.components, routerState.params, sagaMiddleware)
        .then(() => getMarkup(store))
        .then(html => res.status(200).end(html))
        .catch(err => res.end(err.message));
    }
  }));
```

Similar to @nickarra, I have gotten the server-side rendering to work, without the component toggling between loading/loaded states once the client app initializes. Right now I’ve only got one component that extends the WpQuery HLC, so we’ll see how this holds up when I start including additional components.


Cheers,
Ben

—

Ben Tudor
Senior Design Technologist

frog
413 Pine Street, 2nd floor
Seattle, WA 98101

+ 1 206 876 3900 Tel
+ 1 206 876 3901 Fax
ben.tudor@frogdesign.com<mailto:ben.tudor@frogdesign.com>
http://www.frogdesign.com/

design mind: http://designmind.frogdesign.com
twitter: http://twitter.com/frogdesign

**Comment by sdgluck** [comment]
@bt-frog Sorry, your original example was correct. The props object should be passed to the function returned by calling `makePreloader`, not passed to `makePreloader` itself.

**Comment by nickarora** [comment]
@sdgluck @bt-frog sorry for the late response, I was a bit caught up at work this week.

> Has the change you shared meant that server-side rendering is working for you now?

Yes. it is working currently for me; I'm using a forked version of Kasia with the change I described above.  I'll see if I can put together a sample repo to demonstrate the bug I'm describing.

> However there is the whole "not being able to rely on the order of object properties" thing, so an array that tracks all query IDs would be safer.

Completely agree; that seems like an appropriate refactor

> Would you mind sharing what your fetchComponentData (or whatever you're calling it on the server) looks like?

Sure:

```
const renderReact = (renderProps) => {
  const { components } = renderProps

  const sagas = components
      .filter(component => has(component, 'makePreloader')) // check if makePreloader is a key
      .map(component => component.makePreloader()(renderProps))

  // Create a new Redux store instance
  const store = configureStore({})

  const renderedPagePromise = new Promise((resolve) => {
    store.runSaga(waitAll(sagas)).done.then(() => {
      const html = renderToString(
        <Provider store={store}>
          <RouterContext {...renderProps} />
        </Provider>
      )

      // Grab the initial state from our Redux store
      const initialState = store.getState()

      // Send the rendered page back to the client
      resolve(renderFullPage(html, styles, initialState))
    })
  })

  // essentially calls: store.dispatch(END)
  store.close()

  return renderedPagePromise
}
```

**Comment by sdgluck** [comment]
Fixed in v3, see #27 for more details. Thanks for reporting this @nickarora!

**Comment by sdgluck** [comment]
@nickarora @bt-frog I have completed [v1 of the boilerplate](https://github.com/outlandishideas/kasia-boilerplate), maybe it will be useful for you. :-)
